### DISCUSSION TITLE
a simple Blue-berry Pi recipe

### DISCUSSION THREAD

**Discussion by artmg** [snapcast/snapcast#1038] [Show and tell]
# a simple Blue-berry Pi recipe

I have tried several different solutions for a **Bluetooth A2DP input** into snapserver on a **Raspberry Pi** in the years I have been enjoying this fantastic project, but I was never content, because they didn't always 'just work' or they were complex to build and often difficult to diagnose. And there were so many bewildering options to choose from. I am going to share the solution I have finally settled on, after digging in to understand what's going on. I am going for the path of least resistance, and trying to pick the **simplest**, **most lightweight**, **easiest to install** solution that can **still be hacked** if you need.

## Solution overview

My backend audio input solution uses the following layers:

* Operating System: **Raspberry Pi OS Lite**
  * reliable and efficient
  * the **Bluez** Linux Bluetooth subsystem is built-in and configured for the internal radio
* Sound Driver: **ALSA**
  - it comes ready installed, is straight-forward, and it is rarely considered bloated
  - I'm not using the Sound Mixer PulseAudio as snapcast has plenty of control and switching options itself
* Bluetooth Audio API: **BlueALSA**
  * formerly known as Bluez-Alsa this is a lightweight yet mature and widely trusted project
  * it was available out of the box with earlier Pi OS/Raspbian versions
  * the new bluez-alsa-utils package is available from Debian Bookworm version
    * and can be installed onto Bullseye reasonably simply
* **Bluetooth A2DP Player**: bluealsa-aplay
	* comes with BlueALSA - I just package it as a service that runs on connection
	* it can send the sound directly into the `plughw` device on the ALSA loopback below
* **Bluetooth Open Pairing Service**: derived from Bluez sample
	* this service openly allows connections from any device
	* it is based on the `simple-agent` script in the main Bluez project
	* I just have a couple of tweaks to make it promiscuous and silent
* Snapserver source: **ALSA loopback**
	* the audio distribution server picks up sound directly from an ALSA loopback device using its `alsa` protocol.

My hardware choice is a Pi 3 where I will use a wired Ethernet connection, and leave the Wifi disabled. This frees me up to use the internal bluetooth radio without fear of the interference that some have reported. If you are forced to use wireless networking and DO get interference, it should not be hard to use a USB Bluetooth dongle instead.

This has been tested on a clean Raspberry Pi OS Lite Bullseye release April 4th 2022, with Snapserver 0.26.0 .deb installed from the [Snapcast project releases page](https://github.com/badaix/snapcast/releases/). For testing purposes the setup is in two phases. The first sets the Bluetooth A2DP to come out of the headphone jack. The second puts that through into Snapserver. 

## Scripted install for Bluetooth open A2DP receiver

```sh
#### Bluetooth A2DP sink using bluealsa

# This is based on using the built-in Pi 3 bluetooth radio, but networking with 
# ethernet ONLY, so that the internal wifi radio is disabled and can't interfere

# Pi OS 11 'Bullseye' Lite includes bluetooth and bthelper@hci0 services only

# bluealsa not present in Pi bullseye repos - get it from 'Bookworm Raspbian'
# credit https://www.sigmdel.ca/michel/ha/rpi/bluetooth_in_rpios_02_en.html#bluealsa3
echo "deb http://archive.raspbian.org/raspbian/ bookworm main" | sudo tee /etc/apt/sources.list.d/armbian.list
printf 'Package: *\nPin: release n=bookworm\nPin-Priority: 100\n' | sudo tee --append /etc/apt/preferences.d/limit-bookworm
sudo apt update
sudo apt install -y bluez-alsa-utils


##### options for newly-installed bluez-alsa.service

sudo tee /etc/default/bluez-alsa <<EOF!
# Config file for blues-alsa.service - set the service runtime options
# credit https://www.sigmdel.ca/michel/ha/rpi/bluetooth_in_rpios_02_en.html#bluealsa3
#OPTIONS="-p a2dp-source -p a2dp-sink"
# but we only need the sink
OPTIONS="--profile=a2dp-sink"
# although using a USB Bluetooth radio you might need "--device=hciX"
# help - https://github.com/Arkq/bluez-alsa/blob/master/doc/bluealsa.8.rst
EOF!
sudo systemctl restart bluez-alsa.service
sudo systemctl status bluez-alsa.service


# These sections are derived from 
# https://github.com/artmg/rpi-audio-receiver/blob/master/install-bluetooth.sh
# but original credit to https://github.com/nicokaiser/rpi-audio-receiver


##### bluealsa-aplay.service starts after bluez-alsa to pass A2DP to Alsa device

sudo tee /etc/systemd/system/bluealsa-aplay.service <<EOF!
[Unit]
Description=BlueALSA player
Requires=bluez-alsa.service
After=bluez-alsa.service
Wants=bluetooth.target sound.target
[Service]
Type=simple
User=root
ExecStartPre=/bin/sleep 2
ExecStart=/usr/bin/bluealsa-aplay --pcm=default --pcm-buffer-time=250000 00:00:00:00:00:00
# help - https://github.com/Arkq/bluez-alsa/blob/master/doc/bluealsa-aplay.1.rst
[Install]
WantedBy=graphical.target
EOF!

##### udev rule restarts the A2DP player when a device connects to bluetooth

sudo tee /etc/udev/rules.d/61-bluealsa-aplay.rules <<EOF!
# syntax generated by connecting to bluetooth whilst using 
#udevadm monitor --environment --udev --kernel
ACTION=="add", SUBSYSTEM=="bluetooth", RUN+="/bin/systemctl restart bluealsa-aplay.service"
EOF!
# not bothered with udev script for connect/disconnect sounds from original


##### Bluetooth settings

sudo tee /etc/bluetooth/main.conf <<EOF!
[General]
# With bluetooth Class of Device (CoD), the Service Class (>0x2000) might be ignored
#Class = 0x200414
# Major device class: Audio/Video; Minor device class: Hifi audio device;
#Class = 0x0428
# Major device class: Audio/Video; Minor device class: Loudspeaker;
Class = 0x0414
DiscoverableTimeout = 0
[Policy]
AutoEnable=true
EOF!

sudo service bluetooth restart
sudo hciconfig hci0 piscan
sudo hciconfig hci0 sspmode 1

##### ALSA settings

sudo sed -i.orig 's/^options snd-usb-audio index=-2$/#options snd-usb-audio index=-2/' /lib/modprobe.d/aliases.conf


##### Bluetooth open pairing agent

# use modified version of BlueZ 'simple agent' python test script
wget https://raw.githubusercontent.com/bluez/bluez/master/test/simple-agent
# credit - https://github.com/RPi-Distro/repo/issues/291#issuecomment-1149427137
sed -i.orig 's/capability = "KeyboardDisplay"/capability = "NoInputNoOutput"/' simple-agent
sed -i.orig 's/return raw_input(prompt)/return "yes"/' simple-agent
sudo cp simple-agent /usr/local/bin/bluetooth-open-pairing
sudo chmod 755 /usr/local/bin/bluetooth-open-pairing
# and required library script
wget https://raw.githubusercontent.com/bluez/bluez/master/test/bluezutils.py
sudo cp bluezutils.py /usr/local/bin/
sudo chmod 755 /usr/local/bin/bluezutils.py
# I chose this script as it comes from the Bluez project itself. 
# If it ever changes and you can't make it work, consider a simpler derivative like 
# https://gist.github.com/mill1000/74c7473ee3b4a5b13f6325e9994ff84c#file-a2dp-agent

sudo tee /etc/systemd/system/bluetooth-open-pairing.service <<EOF!
# runs a modified copy of bluez simple-agent to silently allow all pairing requests
[Unit]
Description=Bluetooth open pairing agent
Requires=bluetooth.service
After=bluetooth.service
[Install]
WantedBy=bluetooth.target
[Service]
Type=simple
ExecStartPre=/bin/hciconfig hci0 piscan
ExecStartPre=/bin/hciconfig hci0 sspmode 1
ExecStart=/usr/local/bin/bluetooth-open-pairing
Restart=always
RestartSec=10
EOF!

# install pairing agent script dependencies 
sudo apt install -y python3-gi python3-dbus

# now let's run it all
sudo systemctl daemon-reload
sudo systemctl enable --now bluealsa-aplay
sudo systemctl enable --now bluetooth-open-pairing.service

# Now you should be able to pair with your Pi and send sound 
# that you can hear when you plug into the internal headphone jack

```

## Send A2DP stream to Snapserver

```sh
sudo tee -a /etc/modules <<EOF!
# add alsa loopback module for bluealsa-aplay to send bluetooth to snapcast
snd-aloop
EOF!


# direct the bluealsa sound into the loopback device

sudo sed -i.orig 's/--pcm=default/--pcm=plughw:0,0,0/' /etc/systemd/system/bluealsa-aplay.service


### Now set the source(s) you want in snapserver

sudo tee /etc/snapserver.conf <<EOF!
# help - https://github.com/badaix/snapcast/blob/master/doc/player_setup.md
# and - https://github.com/badaix/snapcast/blob/master/doc/configuration.md

# Name is what will appear when choosing the stream using snapcast API

[stream]
# when you send to device 0,0,0 it will loop back on the next device 0,1,0
# to hear a bluetooth stream appears to need a silence trigger
source = alsa://?name=Bluetooth&device=hw:0,1,0&silence_threshold_percent=0.1
EOF!


### and reboot ready to Test it out

sudo reboot 0
# you need to restart so the loopback device gets card #0
```

#### diagnostics

```sh
# list the alsa devices
aplay -l
# list the devices connected through bluetooth
bluealsa-aplay -l

service bluealsa-aplay status
service snapserver status

# if you prefer to turn the pager off...
SYSTEMD_PAGER=

systemctl status bluetooth.service
systemctl status bluez-alsa.service
systemctl status bluealsa-aplay.service
systemctl status bluetooth-open-pairing.service

sudo service snapserver stop
snapserver --logging.filter debug

```

More background and tweaks over time can be found in my source documents on the [open Bluetooth reciever](https://github.com/artmg/MuGammaPi/wiki/Bluetooth-audio-sink#scripted-install-for-bluetooth-open-a2dp-receiver) and the [A2DP to Snapserver link](https://github.com/artmg/MuGammaPi/wiki/multi-room-audio#bluetooth-sink).

I welcome comments and questions.
Enjoy!

**Comment by kingosticks** [comment]
Did you see the bluealsa-aplay option to play from only one device at a time? That is probably what most people would want: `--single-audio`

**Reply by artmg** [reply]
Ahh, I had not been 'foolish' enough to attempt multiple concurrent connections and streams. Many thanks for the tip!
